### PR TITLE
[2/4] Add dig.Out type

### PR DESCRIPTION
**Pull request by glibsm**
Second step of #80 

Needs basic README docs, but so does `dig.In` usage. Will work on a separate PR for it

### CODE REVIEW

**Comment by glibsm** [comment]
There are a lot of similarities for how we traverse over all nested objects in `dig.In` and `dig.Out`, yet they use completely different code paths.

There are also a lot of similarities for when we traverse over just types vs when we go over types *AND* values after a constructor has been called, there is probably some DRY-ness there too.

I will see if all four cases can be united, because the only things that differ are:
- Conditions for recursion (bool function would work)
- Types only, or values are also present

File: stringer.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	fmt.Fprintln(b, "}")
+
+	return b.String()
+}
+
+func (n node) String() string {
+	return fmt.Sprintf(
+		"deps: %v, constructor: %v", n.deps, n.ctype,
+	)
+}
```
Comment (glibsm) [inline stringer.go]: I will change the base branch, or merge this in first so it goes away from this PR

**Comment by glibsm** [comment]
Dropping the WIP tag, as it's now fully working e2e and I was able to refactor the traverse

File: dig.go [hunk 2]
```diff
@@ -172,6 +150,77 @@ func (c *Container) provideConstructor(ctor interface{}, ctype reflect.Type) err
...
+				continue // skip private fields
+			}
+
+			// keep recursing to traverse all the embedded objects
+			traverseOutTypes(ft, f)
+		}
+	} else {
+		// call the provided function on non-Out type
+		if err := f(t); err != nil {
+			return err
```
Comment (abhinav) [inline dig.go]: nit: invert condition

```
if !isOutObject(t) {
  return f(t)
}

// ..
```

File: dig.go [hunk 3]
```diff
@@ -210,14 +259,68 @@ func (c *Container) get(t reflect.Type) (reflect.Value, error) {
...
+				fv    = v.Field(i)
+			)
+			// recurse into all embedded objects
+			c.set(ft, fv)
+		}
+	} else {
+		// do not cache error types
+		if t != _errType {
+			c.cache[t] = v
+		}
```
Comment (abhinav) [inline dig.go]: nit: invert if condition to reduce nesting

```go
if !isOutObject(t) {
  if t != _errType {
    c.cache[t] = v
  }
  return
}

// ..
```

File: dig.go [hunk 4]
```diff
@@ -210,14 +259,69 @@ func (c *Container) get(t reflect.Type) (reflect.Value, error) {
 
 	for _, con := range constructed {
 		ct := con.Type()
-		if ct == _errType {
-			continue
-		}
-		c.cache[ct] = con
+		c.set(ct, con)
```
Comment (akshayjshah) [inline dig.go]: Can `set` take just `con` and handle getting the type internally? Prevents unintentional mismatches between the passed value and type.

File: dig.go [hunk 5]
```diff
@@ -210,14 +259,69 @@ func (c *Container) get(t reflect.Type) (reflect.Value, error) {
...
+func (c *Container) set(t reflect.Type, v reflect.Value) {
+	if !isOutObject(t) {
+		// do not cache error types
+		if t != _errType {
+			c.cache[t] = v
+		}
+		return
+	}
+
+	// dig.Out objects are not acted upon directly, but rather their memebers are considered
```
Comment (akshayjshah) [inline dig.go]: spelling: members

File: dig.go [hunk 6]
```diff
@@ -210,14 +259,69 @@ func (c *Container) get(t reflect.Type) (reflect.Value, error) {
...
+	for i := 0; i < t.NumField(); i++ {
+		var (
+			field = t.Field(i)
+			ft    = field.Type
+			fv    = v.Field(i)
+		)
+		// recurse into all embedded objects
```
Comment (akshayjshah) [inline dig.go]: nit: these are fields, but they're not necessarily embedded. https://golang.org/ref/spec#Struct_types

File: dig_test.go [hunk 7]
```diff
@@ -427,6 +427,31 @@ func TestEndToEndSuccess(t *testing.T) {
...
+		}
+		type B struct {
+			name string
+		}
+		type Ret struct {
+			Out
+
+			foo string // private field to be ignored
+			A          // value type A
+			*B         // pointer type *B
```
Comment (akshayjshah) [inline dig_test.go]: nit: embeds at the top.

File: dig_test.go [hunk 8]
```diff
@@ -427,6 +427,31 @@ func TestEndToEndSuccess(t *testing.T) {
...
+		}
+
+		c := New()
+		require.NoError(t, c.Provide(func() Ret {
+			return Ret{A: A{"string A"}, B: &B{"string B"}}
+		}), "provide for the Ret struct should succeed")
+		require.NoError(t, c.Invoke(func(a A, b *B) {
+			assert.Equal(t, a.name, "string A", "value type should work for dig.Out")
+			assert.Equal(t, b.name, "string B", "pointer sohuld work for dig.Out")
```
Comment (akshayjshah) [inline dig_test.go]: spelling: "should"

Also, can we be nit-picky and also validate that we're getting the same ref returned from the constructors?

File: dig.go (hunk 5 above)
Comment (glibsm) [inline dig.go]: memebears!

**Pull request merged**